codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java

```diff
@@ -555,33 +585,44 @@ final void weight(short weight) {
          */
         final void takeChild(DefaultStream child, boolean exclusive, List<ParentChangedEvent> events) {
             DefaultStream oldParent = child.parent();
-            events.add(new ParentChangedEvent(child, oldParent));
-            notifyParentChanging(child, this);
-            child.parent = this;
 
-            if (exclusive && !children.isEmpty()) {
-                // If it was requested that this child be the exclusive dependency of this node,
-                // move any previous children to the child node, becoming grand children of this node.
-                for (DefaultStream grandchild : removeAllChildren().values()) {
-                    child.takeChild(grandchild, false, events);
+            if (oldParent != this) {
+                events.add(new ParentChangedEvent(child, oldParent));
+                notifyParentChanging(child, this);
+                child.parent = this;
+                // We need the removal operation to happen first so the prioritizableForTree for the old parent to root
+                // path is updated with the correct child.prioritizableForTree() value. Note that the removal operation
+                // may not be successful and may return null. This is because when an exclusive dependency is processed
+                // the children are removed in a previous recursive call but the child's parent link is updated here.
+                if (oldParent != null && oldParent.children.remove(child.id()) != null) {
```

Comment:
@nmittler - Turns out it is possible (and legit with the current code structure) for `oldParent.children.remove(child.id()) == null`.

> @nmittler - The `retain(..)` call when an exclusive dependency happens is messing things up.  We remove from the children collection in a special way such that we don't update the child's parent link until the next `takeChild` recursive call. This means that the `oldParent` won't have the child in its `children` collection...